pkg/brb/brbmodule.go

```diff
@@ -0,0 +1,202 @@
+	MessageContent
+	int
+}) {
+	// TODO cant store []byte in a map
+	var stringRepr = string(*message)
+
+	(*counts)[stringRepr]++
+
+	if (*counts)[stringRepr] > (*accumulator).int {
+		(*accumulator).int = (*counts)[stringRepr]
+		(*accumulator).MessageContent = *message
+	}
+}
+
+func NewModule(mc *ModuleConfig, params *ModuleParams, nodeID t.NodeID) modules.PassiveModule {
+	m := dsl.NewModule(mc.Self)
+
+	// upon event <brb, Init> do
+	state := make(map[int64]*brbModuleState)
+	lastId := int64(0)
+
+	// upon event <brb, Broadcast | m> do
+	brbpbdsl.UponBroadcastRequest(m, func(id int64, data []byte) error {
+		if id <= lastId {
+			return nil
+		}
+		if nodeID != params.Leader {
+			return fmt.Errorf("only the leader node can receive requests")
+		}
+		initialize(&state, id)
+		state[id].request = data
+		eventpbdsl.SendMessage(m, mc.Net, brbpbmsgs.StartMessage(mc.Self, id, data), params.AllNodes)
+		return nil
+	})
+
+	brbpbdsl.UponStartMessageReceived(m, func(from t.NodeID, id int64, data []byte) error {
+		if id <= lastId {
+			return nil
+		}
+		if from == params.Leader {
+			initialize(&state, id)
+			if state[id].sentEcho == false {
+				state[id].sentEcho = true
+				eventpbdsl.SendMessage(m, mc.Net, brbpbmsgs.EchoMessage(mc.Self, id, data), params.AllNodes)
+				return nil
+			} else {
+				return fmt.Errorf("already sent echo")
+			}
+		} else {
+			return fmt.Errorf("received start message not from leader")
+		}
+	})
+
+	brbpbdsl.UponEchoMessageReceived(m, func(from t.NodeID, id int64, data []byte) error {
+		if id <= lastId {
+			return nil
+		}
+		initialize(&state, id)
+		if state[id].echos[from] == nil {
+			state[id].echos[from] = data
+
+			incrementAndUpdateAccumulator(&data, &state[id].echoMessagesCount, &state[id].echosMaxAccumulator)
+		}
+		return nil
+	})
+
+	brbpbdsl.UponReadyMessageReceived(m, func(from t.NodeID, id int64, data []byte) error {
+		if id <= lastId {
+			return nil
+		}
+		initialize(&state, id)
+		if state[id].readys[from] == nil {
+			state[id].readys[from] = data
+
+			incrementAndUpdateAccumulator(&data, &state[id].readyMessagesCount, &state[id].readyMaxAccumulator)
+		}
+		return nil
+	})
+
+	dsl.UponCondition(m, func() error {
+		for id, currentState := range state {
+			if id <= lastId {
+				continue
+			}
+			if (currentState.echosMaxAccumulator.int > (params.GetN()+params.GetF())/2 || currentState.readyMaxAccumulator.int > params.GetF()) && currentState.sentReady == false {
+				currentState.sentReady = true
+				eventpbdsl.SendMessage(m, mc.Net, brbpbmsgs.ReadyMessage(mc.Self, id, currentState.echosMaxAccumulator.MessageContent), params.AllNodes)
+			}
+
+			if currentState.readyMaxAccumulator.int > 2*params.GetF() && currentState.delivered == false {
+				currentState.delivered = true
+				if id > lastId {
+					lastId = id
+				}
+				brbpbdsl.Deliver(m, mc.Consumer, id, currentState.readyMaxAccumulator.MessageContent)
+				delete(state, id)
```

Comment:
Note that you may "jump" over some IDs, thus, essentially, leaving memory leaks behind.